### PR TITLE
[dashboards] Formula and Function support for Query Value Widgets in Dashboard resource

### PR DESCRIPTION
**Pull request by DrkSephy**
This pull request implements support for formulas and functions in Query Value widgets.

Note: this is not generally available yet.

Related PRs: 

Merged: [[dashboards] Formula and Function support for Timeseries Widgets in Dashboard resource](https://github.com/DataDog/terraform-provider-datadog/pull/892)

[[dashboards] Formula and Function support for Toplist Widgets in Dashboard resource](https://github.com/DataDog/terraform-provider-datadog/pull/951)

### CODE REVIEW

File: datadog/resource_datadog_dashboard.go [hunk 1]
```diff
@@ -4901,8 +4932,10 @@ func buildDatadogEventQuery(data map[string]interface{}) datadogV1.FormulaAndFun
 	eventQuery.SetIndexes(indexes)
 
 	if terraformSearches, ok := data["search"].([]interface{}); ok && len(terraformSearches) > 0 {
-		terraformSearch := terraformSearches[0].(map[string]interface{})
-		eventQuery.Search = datadogV1.NewFormulaAndFunctionEventQueryDefinitionSearch(terraformSearch["query"].(string))
+		terraformSearch, ok := terraformSearches[0].(map[string]interface{})
+		if ok {
+			eventQuery.Search = datadogV1.NewFormulaAndFunctionEventQueryDefinitionSearch(terraformSearch["query"].(string))
+		}
```
Comment (DrkSephy) [inline datadog/resource_datadog_dashboard.go:4938]: This prevents a crash when `search.query = ""`, but it raises the same issue mentioned [here](https://github.com/DataDog/terraform-provider-datadog/pull/951/files#r582418101).

Comment (DrkSephy) [reply datadog/resource_datadog_dashboard.go]: Resolved this by using a newly build string validator in [this commit](https://github.com/DataDog/terraform-provider-datadog/pull/953/commits/4a225543834ef2aa85a9579eba380986c580a915).

**Review by matt-miller-ddog — APPROVED** [review]

**Review by therve — APPROVED** [review]

**Review by zippolyte — APPROVED** [review]
LGTM

**Pull request merged**